src/metabase/models/setting.clj

```diff
@@ -1342,3 +1354,57 @@
            (map (fn [[setting-name]]
                   [setting-name (get setting-name)])))
      @registered-settings)))
+
+(defn- redact-parse-ex
+  "Substitute an opaque exception to ensure no sensitive information in the raw value is exposed"
+  [ex]
+  (ex-info (trs "Error of type {0} thrown while parsing a setting" (type ex))
+           {:ex-type (type ex)}))
+
+(defn- may-contain-raw-token? [ex setting]
+  (case (:type setting)
+    :json
+    (cond
+      (instance? JsonEOFException ex) false
+      (instance? JsonParseException ex) true
+      :else (do (log/warn ex "Unexpected exception while parsing JSON")
+                ;; err on the side of caution
+                true))
+
+    ;; TODO: handle the remaining formats explicitly
+    true))
+
+(defn- redact-sensitive-tokens [ex raw-value]
+  (if (may-contain-raw-token? ex raw-value)
+    (redact-parse-ex ex)
+    ex))
+
+(defn- validate-setting
+  "Test whether the value configured for a given setting can be parsed as the expected type.
+   Returns an map containing the exception if an issue is encountered, or nil if the value passes validation."
+  [setting]
+  (try
+    (get-value-of-type (:type setting) setting)
+    nil
+    (catch java.lang.Exception e
+      (let [parse-error (ex-cause e)
```

Comment:
Not every exception is guaranteed to have a cause. So `parse-error` can be `nil`. I think we'll need to handle each type of exception separately, sometimes we'll want to display the message, sometimes the cause. Using the cause would be fine if we're just talking about JSON parsing exceptions, but I don't think this generalizes to all types of setting parsing exceptions.

I noticed there are also no tests for other types of settings. What about limiting this check to settings of type `:json` for now? If so, the rest of the code looks good.

Reply:
Ah you're right, there will always be a cause. It's not clear from the code these things are coupled like that. I think 2 helps decouple them a little, as for 1, I'm not too sure what it achieves. Maybe there's a better way to be explicit about coupling, but I don't think it's so important if the fallback is there and we have tests. There should still be tests for other types of settings though.